src/Nethermind/Nethermind.Evm/StackAccessTracker.cs

```diff
@@ -1,102 +1,126 @@
+                foreach (UInt256 storage in storages)
                 {
-                    _accessedAddresses.Add(address);
-                    foreach (UInt256 storage in storages)
-                    {
-                        _accessedStorageCells.Add(new StorageCell(address, storage));
-                    }
+                    _trackingState.AccessedStorageCells.Add(new StorageCell(address, storage));
                 }
             }
         }
+    }
 
-        public readonly void ToBeDestroyed(Address address)
-        {
-            _destroyList.Add(address);
-        }
+    public readonly void ToBeDestroyed(Address address)
+    {
+        _trackingState.DestroyList.Add(address);
+    }
 
-        public readonly void WasCreated(Address address)
-        {
-            _createList.Add(address);
-        }
+    public readonly void WasCreated(Address address)
+    {
+        _trackingState.CreateList.Add(address);
+    }
+
+    public void TakeSnapshot()
+    {
+        _addressesSnapshots = _trackingState.AccessedAddresses.TakeSnapshot();
+        _storageKeysSnapshots = _trackingState.AccessedStorageCells.TakeSnapshot();
+        _destroyListSnapshots = _trackingState.DestroyList.TakeSnapshot();
+        _logsSnapshots = _trackingState.Logs.TakeSnapshot();
+    }
+
+    public readonly void Restore()
+    {
+        _trackingState.Logs.Restore(_logsSnapshots);
+        _trackingState.DestroyList.Restore(_destroyListSnapshots);
+        _trackingState.AccessedAddresses.Restore(_addressesSnapshots);
+        _trackingState.AccessedStorageCells.Restore(_storageKeysSnapshots);
+    }
+
+    public void Dispose()
+    {
+        TrackingState state = _trackingState;
+        _trackingState = null;
+        TrackingState.ResetAndReturn(state);
+    }
 
-        public void TakeSnapshot()
+    private class TrackingState
+    {
+        private readonly static ConcurrentQueue<TrackingState> _trackerPool = new();
+        public static TrackingState RentState() => _trackerPool.TryDequeue(out TrackingState tracker) ? tracker : new TrackingState();
+
+        public static void ResetAndReturn(TrackingState state)
         {
-            _addressesSnapshots = _accessedAddresses.TakeSnapshot();
-            _storageKeysSnapshots = _accessedStorageCells.TakeSnapshot();
-            _destroyListSnapshots = _destroyList.TakeSnapshot();
-            _logsSnapshots = _logs.TakeSnapshot();
+            state.Clear();
+            _trackerPool.Enqueue(state);
         }
 
-        public readonly void Restore()
+        public JournalSet<Address> AccessedAddresses { get; } = new();
+        public JournalSet<StorageCell> AccessedStorageCells { get; } = new();
+        public JournalCollection<LogEntry> Logs { get; } = new();
+        public JournalSet<Address> DestroyList { get; } = new();
+        public HashSet<AddressAsKey> CreateList { get; } = new();
+
+        public void Clear()
```

Comment:
This method can be inlined in `ResetAndReturn`

Reply:
Static method vs instance?

Reply:
ok at least it doesn't need to be public

Reply:
Made private